cabal-install/Distribution/Client/Init.hs

```diff
@@ -677,6 +678,38 @@ createSourceDirectories flags = case sourceDirs flags of
                                   Just dirs -> forM_ dirs (createDirectoryIfMissing True)
                                   Nothing   -> return ()
 
+-- | Create Main.hs, but only if we are init'ing an executable and
+--   the mainIs flag has been provided.
+createMainHs :: InitFlags -> IO ()
+createMainHs flags@InitFlags{ sourceDirs = Just (srcPath:_)
+                            , packageType = Flag Executable
+                            , mainIs = Flag mainFile } =
+  writeMainHs flags (srcPath </> mainFile)
+createMainHs flags@InitFlags{ sourceDirs = _
+                            , packageType = Flag Executable
+                            , mainIs = Flag mainFile } =
+  writeMainHs flags mainFile
+createMainHs _ = return ()
+
+-- | Write a main file if it doesn't already exist.
+writeMainHs :: InitFlags -> FilePath -> IO ()
+writeMainHs flags mainPath = do
+  dir <- maybe getCurrentDirectory return (flagToMaybe $ packageDir flags)
+  let mainFullPath = dir </> mainPath
+  exists <- doesFileExist mainFullPath
+  unless exists $ do
+      message flags $ "Generating " ++ mainPath ++ "..."
+      writeFileSafe flags mainFullPath mainHs
+
+-- | Default Main.hs file.  Used when no Main.hs exists.
+mainHs :: String
+mainHs = unlines
+  [ "module Main where"
+  , ""
+  , "main :: IO ()"
+  , "main = putStrLn \"Hello, World!\""
```

Comment:
This sounds great!  I'd appreciate some bikeshedding on what it should say.  (I don't particularly have any good ideas on what it should say!)